libbeat/publisher/queue/qos.go

```diff
@@ -0,0 +1,112 @@
+package queue
+
+import (
+	"container/heap"
+	"time"
+
+	"github.com/elastic/beats/libbeat/common"
+)
+
+type Qoser interface {
+	// Release the process lock and give the processing time to the next available producer
+	ReleaseToken()
+
+	// Schedule picks the next producer who gets to produce to the queue
+	Schedule()
+
+	// Clone an instance of QoSer
+	CreateClient() QoserClient
+}
+
+type QoserClient interface {
+	// Add the producer in the wait queue based
+	RequestToken(*ProducerWrapper, int, time.Time)
+
+	// Wait until the next available slot
+	Wait()
+
+	// Send a signal to the producer to wake up
+	Awaken()
+}
+
+type weightedScheduler struct {
+	heap heap.Interface
+
+	notify chan struct{}
+}
+
+func NewWeightedScheduler() Qoser {
+	return &weightedScheduler{
+		heap:   NewHeap(),
+		notify: make(chan struct{}),
+	}
+}
+
+func (w *weightedScheduler) ReleaseToken() {
+	// Pop the most high priority producer
+	iface := w.heap.Pop()
+
+	if iface == nil {
+		return
+	}
+
+	if it, ok := iface.(*item); ok {
+		producer := it.value.(*ProducerWrapper)
+		// Notify the producer to wake up.
+		producer.Awaken()
+	}
+}
+
+func (w *weightedScheduler) Schedule() {
+	go func() {
+		for {
+			select {
+			case <-w.notify:
+				w.ReleaseToken()
```

Comment:
I wonder how effective this scheduler is. I'm a little afraid the heap will be of length 1 most of the time. The scheduler shown here very much relies on contention on the heap structure.